app/src/hooks/activity-value-form/use-emission-factors.ts

```diff
@@ -0,0 +1,155 @@
+// take the methodology selected,
+
+import { api } from "@/services/api";
+import { ExtraField } from "@/util/form-schema";
+import { useEffect, useMemo } from "react";
+import { Control, UseFormSetValue, useWatch } from "react-hook-form";
+import { EmissionsFactorResponse } from "@/util/types";
+import { getTranslationFromDict } from "@/i18n";
+import { uniqBy } from "lodash";
+
+const reduceEmissionsToUniqueSourcesAndUnits = (
+  emissionsFactors: EmissionsFactorResponse,
+) => {
+  const reducedMap: {
+    [key: string]: {
+      id: string;
+      name: string;
+      gasValuesByGas: {
+        [gas: string]: {
+          gasValues: Record<string, any>[];
+        };
+      };
+    };
+  } = {};
+
+  emissionsFactors.forEach((factor) => {
+    factor.dataSources.forEach((source) => {
+      // Initialize the data source if it doesn't exist in reducedMap
+      if (!reducedMap[source.datasourceId]) {
+        reducedMap[source.datasourceId] = {
+          id: source.datasourceId,
+          name: getTranslationFromDict(source.datasetName) ?? "unknown",
+          gasValuesByGas: {},
+        };
+      }
+
+      if (factor.gas) {
+        if (!reducedMap[source.datasourceId].gasValuesByGas[factor.gas]) {
+          reducedMap[source.datasourceId].gasValuesByGas[factor.gas] = {
+            gasValues: [],
+          };
+        }
+
+        reducedMap[source.datasourceId].gasValuesByGas[
+          factor.gas
+        ].gasValues.push({
+          ...factor,
+          gas: factor.gas as string,
+          emissionsPerActivity: factor.emissionsPerActivity as number,
+          datasource: null,
+        });
+      }
+    });
+  });
+
+  // Return the array of unique data sources with gas values grouped by units
+  return Object.values(reducedMap).map((source) => ({
+    ...source,
+    gasValuesByGas: Object.keys(source.gasValuesByGas).reduce(
+      (acc, currentValue) => {
+        return {
+          ...acc,
+          [currentValue]: {
+            gasValues: uniqBy(
+              source.gasValuesByGas[currentValue].gasValues,
+              "emissionsPerActivity",
+            ).filter((factor) => ["kg/m3", "kg/kWh"].includes(factor.units)), // filter only emissions that have kg/m3 or kg/kWh as the unit
```

Comment:
Why only these?

Reply:
These are the units that every emission factor should be in. I and @mfonsecaOEF  agreed that the emissions factors should all be in the same units. 

@mfonsecaOEF  is still working on converting all EFs to these units. But until she's done,  i'm filtering out EFs that are not these units so that the calculations will be accurate